src/shared/bootspec.c

```diff
@@ -1392,13 +1392,141 @@ static void boot_entry_file_list(
                 *ret_status = status;
 }
 
+static int copy_cmdline(
+                char *cmdline,
+                char **ret) {
+
+        assert(ret);
+
+        _cleanup_free_ char *temp = *ret;
+
+        *ret = strnappend(temp, cmdline, strlen(cmdline));
+        if (temp)
+                free(temp);
+        temp = *ret;
+
+        *ret = strnappend(temp, " ", 1);
+        if (!*ret)
+                return log_oom();
+
+        return 0;
+}
+
+static void print_addon(
+                BootEntryAddon *addon,
+                const char *addon_str) {
+
+        printf("  %s: %s\n", addon_str, addon->location);
+        printf("      cmdline: %s%s\n", special_glyph(SPECIAL_GLYPH_TREE_RIGHT), addon->cmdline);
+}
+
+static int print_cmdline(const BootEntry *e) {
+
+        _cleanup_free_ char *final_cmdline = NULL;
+        int r;
+
+        assert(e);
+
+        if (!strv_isempty(e->options)) {
+                _cleanup_free_ char *t = NULL, *t2 = NULL;
+                _cleanup_strv_free_ char **ts = NULL;
+
+                t = strv_join(e->options, " ");
+                if (!t)
+                        return log_oom();
+
+                ts = strv_split_newlines(t);
+                if (!ts)
+                        return log_oom();
+
+                t2 = strv_join(ts, "\n              ");
+                if (!t2)
+                        return log_oom();
+
+                printf("  defCmdline: %s\n", t2);
+                r = copy_cmdline(t2, &final_cmdline);
+                if (r < 0)
+                        return r;
+        }
+
+        LIST_FOREACH(addon_list, addon, e->local_addon_list) {
+                /* Add space at the beginning of addon_str to align it correctly */
+                print_addon(addon, " localAddon");
+                r = copy_cmdline(addon->cmdline, &final_cmdline);
+                if (r < 0)
+                        return r;
+        }
+
+        if (final_cmdline)
+                printf(" finalCmdline: %s\n", final_cmdline);
+
+        return 0;
+}
+
+static int json_addon(
+                BootEntryAddon *addon,
+                const char *addon_str,
+                JsonVariant **array) {
+
+        _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+        int r;
+
+        r = json_variant_merge_objectb(
+                        &v, JSON_BUILD_OBJECT(
+                                JSON_BUILD_PAIR(addon_str, JSON_BUILD_STRING(addon->location)),
+                                JSON_BUILD_PAIR("cmdline", JSON_BUILD_STRING(addon->cmdline))));
+        if (r < 0)
+                return log_oom();
+
+        r = json_variant_append_array(array, v);
```

Comment:
you can just use json_variant_append_arrayb(), which combines json_variant_append_array() with json_build() into one